src/main/java/bio/terra/workspace/service/resource/reference/ReferenceResourceService.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    return getReferenceResource(resource.getWorkspaceId(), resourceIdResult, userReq);
+  }
+
+  /**
+   * At this moment, the only updates on references we are doing are to name and description so this
+   * is a common implementation. If we get more complicated, we can break it out.
+   *
+   * @param workspaceId workspace of interest
+   * @param resourceId resource to update
+   * @param name name to change - may be null
+   * @param description description to change - may be null
+   */
+  public void updateReferenceResource(
+      UUID workspaceId,
+      UUID resourceId,
+      String name,
+      String description,
+      AuthenticatedUserRequest userReq) {
+    workspaceService.validateWorkspaceAndAction(
+        userReq, workspaceId, SamConstants.SAM_WORKSPACE_WRITE_ACTION);
+    resourceDao.updateReferenceResource(workspaceId, resourceId, name, description);
+  }
+
+  /**
+   * Delete a reference. The only state we hold for a reference is in the metadata database so we
+   * directly delete that.
+   *
+   * @param workspaceId workspace of interest
+   * @param resourceId resource to delete
+   * @param userReq authenticated user
+   */
+  public void deleteReferenceResource(
+      UUID workspaceId, UUID resourceId, AuthenticatedUserRequest userReq) {
+    // TODO: Fix this based on resolution of permission model issue.
+    workspaceService.validateWorkspaceAndAction(
+        userReq, workspaceId, SamConstants.SAM_WORKSPACE_WRITE_ACTION);
+    resourceDao.deleteResource(workspaceId, resourceId);
+  }
+
+  public ReferenceResource getReferenceResource(
+      UUID workspaceId, UUID resourceId, AuthenticatedUserRequest userReq) {
+    // TODO: Fix this based on resolution of permission model issue.
+    workspaceService.validateWorkspaceAndAction(
+        userReq, workspaceId, SamConstants.SAM_WORKSPACE_READ_ACTION);
+    WsmResource wsmResource = resourceDao.getResource(workspaceId, resourceId);
+    return castReferenceResource(wsmResource);
+  }
+
+  public ReferenceResource getReferenceResourceByName(
+      UUID workspaceId, String name, AuthenticatedUserRequest userReq) {
+    // TODO: Fix this based on resolution of permission model issue.
+    workspaceService.validateWorkspaceAndAction(
+        userReq, workspaceId, SamConstants.SAM_WORKSPACE_READ_ACTION);
+    WsmResource wsmResource = resourceDao.getResourceByName(workspaceId, name);
+    return castReferenceResource(wsmResource);
+  }
+
+  public List<ReferenceResource> enumerateReferences(
+      UUID workspaceId, int offset, int limit, AuthenticatedUserRequest userReq) {
+    // TODO: Fix this based on resolution of permission model issue.
+    workspaceService.validateWorkspaceAndAction(
+        userReq, workspaceId, SamConstants.SAM_WORKSPACE_READ_ACTION);
+    return resourceDao.enumerateReferences(workspaceId, offset, limit);
+  }
+
+  private ReferenceResource castReferenceResource(WsmResource wsmResource) {
```

Comment:
The other castTo methods don't live in services I thought, but on the hierarchy.

Reply:
Good catch. I'll move them